### PR TITLE
AddCoins

### PR DESCRIPTION
**Pull request by juliankoh**
### Motivation
Adds support for `AddCoins` function that bulk imports coins into the database. 

### Solution
Creates a shared `addCoin` method for adding an individual coin, and a higher level `AddCoins` method for bulk adding coins.

### CODE REVIEW

File: storage/coin_storage.go [hunk 1]
```diff
@@ -97,41 +97,70 @@ func getAndDecodeCoin(
...
 type wallets map[string]map[string]struct{}
 
-func (c *CoinStorage) tryAddingCoin(
+// AccountCoin contains an AccountIdentifier and a Coin that it owns
+type AccountCoin struct {
+	Account *types.AccountIdentifier
+	Coin    *types.Coin
+}
+
+// AddCoins takes an array of AccountIdentifiers and it's respective Coin
```
Comment (patrick-ogrady) [inline storage/coin_storage.go:106]: nit: takes an array of AccountCoin and saves all of them to the database.

File: storage/coin_storage.go [hunk 2]
```diff
@@ -142,21 +171,46 @@ func (c *CoinStorage) tryAddingCoin(
...
 	}
-
-	if _, exists := coins[coinIdentifier.Identifier]; exists {
+	coinIdentifier := coin.CoinIdentifier.Identifier
+	if _, exists := coins[coinIdentifier]; exists {
 		return fmt.Errorf(
 			"coin %s already exists in account %s",
 			coinIdentifier,
-			types.PrettyPrintStruct(operation.Account),
+			types.PrettyPrintStruct(account),
```
Comment (patrick-ogrady) [inline storage/coin_storage.go:179]: nit: `types.PrintStruct` will print an output of the struct without newlines

File: storage/coin_storage_test.go [hunk 3]
```diff
@@ -373,6 +397,45 @@ func TestCoinStorage(t *testing.T) {
...
+				Account: account2,
+				Coin:    coins2,
+			},
+			&AccountCoin{
+				Account: account3,
+				Coin:    coins3,
+			},
+		}
+
+		account1coins, _, _ := c.GetCoins(ctx, account)
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:417]: we should have assertions on these other returns

File: storage/coin_storage_test.go [hunk 4]
```diff
@@ -373,6 +397,45 @@ func TestCoinStorage(t *testing.T) {
...
+		account1coins = append(account1coins, coins1)
+		assert.NoError(t, err)
+		assert.ElementsMatch(t, coinsGot, account1coins)
+
+		coinsGot, _, err = c.GetCoins(ctx, account2)
+		account2coins = append(account2coins, coins2)
+		assert.NoError(t, err)
+		assert.ElementsMatch(t, coinsGot, account2coins)
+
+		coinsGot, _, err = c.GetCoins(ctx, account3)
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:434]: assertion on other returns

File: storage/coin_storage_test.go [hunk 5]
```diff
@@ -373,6 +397,45 @@ func TestCoinStorage(t *testing.T) {
 		assert.Equal(t, &types.CoinIdentifier{Identifier: "coin4"}, coinIdentifier)
 		assert.Equal(t, blockIdentifier, block)
 	})
+
+	t.Run("AddCoins", func(t *testing.T) {
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:401]: This test only tries to add coins after we've added blocks. We should also attempt to add coins before we've processed any blocks.

File: storage/coin_storage_test.go [hunk 6]
```diff
@@ -373,6 +397,45 @@ func TestCoinStorage(t *testing.T) {
...
+		accountCoins := []*AccountCoin{
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:402]: Can we try adding a coin that already exists to make sure we don't put it in twice?

File: storage/coin_storage_test.go [hunk 7]
```diff
@@ -373,6 +447,66 @@ func TestCoinStorage(t *testing.T) {
...
+
+		// Does not add duplicate
+		coinsGot, block, err = c.GetCoins(ctx, account4)
+		assert.NoError(t, err)
+		assert.Equal(t, blockIdentifier, block)
+		assert.ElementsMatch(t, coinsGot, []*types.Coin{coins4})
+	})
 }
 
 type MockCoinStorageHelper struct{}
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:512]: realized my mocking here is sub-par...you should allow for current block identifier to return an ErrHeadBlockNotFound here when adding coins (right now we always return a block as if we had synced)

**Comment by heimdall-asguard** [comment]
Review Error for patrick-ogrady @ 2020-08-19 17:55:08 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

File: storage/coin_storage_test.go [hunk 8]
```diff
@@ -258,6 +306,46 @@ func TestCoinStorage(t *testing.T) {
 
 	c := NewCoinStorage(database, mockHelper, a)
 
+	t.Run("AddCoins before blocks", func(t *testing.T) {
```
Comment (patrick-ogrady) [inline storage/coin_storage_test.go:309]: Thanks for adding this test!

**Pull request merged**